host/windows/syscall.c

```diff
@@ -408,214 +1201,712 @@ ssize_t oe_syscall_readv_ocall(
+    // suppose fd is a handle.
+    HANDLE oldhandle = (HANDLE)fd;
+
+    // If fd is a stdin/out/err, convert it to the corresponding HANDLE.
+    switch (fd)
+    {
+        case OE_STDIN_FILENO:
+            oldhandle = GetStdHandle(STD_INPUT_HANDLE);
+            break;
+
+        case OE_STDOUT_FILENO:
+            oldhandle = GetStdHandle(STD_OUTPUT_HANDLE);
+            break;
+
+        case OE_STDERR_FILENO:
+            oldhandle = GetStdHandle(STD_ERROR_HANDLE);
+            break;
+
+        default:
+            break;
+    }
+
+    // Check if oldhanle is a valid file handle.
+    DWORD file_type = GetFileType(oldhandle);
+    // Distinguish between these two cases:
+    // 1. a "valid" return of FILE_TYPE_UNKNOWN: the input is valid.
+    // 2. a return code due to a calling error: the input is not valid.
+    // While GetLastError returns NO_ERROR, it is case 1.
+    if (file_type != FILE_TYPE_UNKNOWN || GetLastError() == NO_ERROR)
+    {
+        if (DuplicateHandle(GetCurrentProcess(),
+                    oldhandle,
+                    GetCurrentProcess(),
+                    (HANDLE*)&ret,
+                    0,
+                    FALSE,
+                    DUPLICATE_SAME_ACCESS))
+        {
+            _set_errno(0);
+        }
+        else
+        {
+            _set_errno(_winerr_to_errno(GetLastError()));
+        }
+    }
+    else
+    {
+        // The input is a oe_socket_fd_t.
+        ret = _dup_socket(fd);
+        if (ret == -1)
+        {
+            _set_errno(OE_EINVAL);
+        }
+        else
+        {
+            _set_errno(0);
+        }
+    }
+
+done:
+    return ret;
+}
+
+uint64_t oe_syscall_opendir_ocall(const char* pathname)
+{
+    struct WIN_DIR_DATA* pdir = NULL;
+
+    pdir = (struct WIN_DIR_DATA*)calloc(1, sizeof(struct WIN_DIR_DATA));
+    if (!pdir)
+    {
+        _set_errno(OE_ENOMEM);
+        goto done;
+    }
+
+    char* wpathname = oe_syscall_path_to_win(pathname, "/*");
+    pdir->hFind = FindFirstFileA(wpathname, &pdir->FindFileData);
+    if (pdir->hFind == INVALID_HANDLE_VALUE)
+    {
+        free(wpathname);
+        free(pdir);
+        pdir = NULL;
+        goto done;
+    }
+
+    pdir->dir_offs = 0;
+    pdir->pdirpath = wpathname;
+
+done:
+    return (uint64_t)pdir;
+}
+
+int oe_syscall_readdir_ocall(uint64_t dirp, struct oe_dirent* entry)
+{
+    int ret = -1;
+
+    struct WIN_DIR_DATA* pdir = (struct WIN_DIR_DATA*)dirp;
+    int nlen = -1;
+
+    _set_errno(0);
+
+    if (!dirp || !entry)
     {
-        errno = EINVAL;
+        _set_errno(OE_EINVAL);
+        ret = -1;
         goto done;
     }
 
-    /* Handle zero data case. */
-    if (!iov || iovcnt == 0)
+    // oe_syscall_opendir_ocall already called FindFirstFileW which returned the '.' entry.
+    // To preserve the readdir semantics, if oe_syscall_readdir_ocall is passed a valid dirp
+    // with the initial offset of 0, we return '.' directly instead of calling FindNextFileW.
```

Comment:
```suggestion
    // with the initial offset of 0, we return '.' directly instead of calling FindNextFileA.
```

Ditto, to match 1480

Reply:
Fixed, thanks.